File: assignments/ruby/point-mutations/point-mutations_test.rb
require 'minitest/autorun'
require_relative 'dna'

class DNATest < MiniTest::Unit::TestCase

  def test_no_difference_between_empty_strands
    assert_equal 0, DNA.new('').hamming_distance('')
  end

  def test_no_difference_between_identical_strands
    skip
    assert_equal 0, DNA.new('GGACTGA').hamming_distance('GGACTGA')
  end

  def test_complete_hamming_distance_in_small_strand
    skip
    assert_equal 3, DNA.new('ACT').hamming_distance('GGA')
  end

  def test_hamming_distance_in_off_by_one_strand
    skip
    assert_equal 19, DNA.new('GGACGGATTCTGACCTGGACTAATTTTGGGG').hamming_distance('AGGACGGATTCTGACCTGGACTAATTTTGGGG')
  end

  def test_small_hamming_distance_in_middle_somewhere
    skip
    assert_equal 1, DNA.new('GGACG').hamming_distance('GGTCG')
  end

  def test_larger_distance
    skip
    assert_equal 2, DNA.new('ACCAGGG').hamming_distance('ACTATGG')
  end

  def test_ignores_extra_length_on_other_strand_when_longer
    skip
    assert_equal 3, DNA.new('AAACTAGGGG').hamming_distance('AGGCTAGCGGTAGGAC')
  end

  def test_ignores_extra_length_on_original_strand_when_longer
    skip
    assert_equal 5, DNA.new('GACTACGGACAGGGTAGGGAAT').hamming_distance('GACATCGCACACC')
  end

  def test_does_not_actually_shorten_original_strand
    skip
    assert_equal 1, DNA.new('AGACAACAGCCAGCCGCCGGATT').hamming_distance('AGGCAA')
    assert_equal 4, DNA.new('AGACAACAGCCAGCCGCCGGATT').hamming_distance('AGACATCTTTCAGCCGCCGGATTAGGCAA')
    assert_equal 1, DNA.new('AGACAACAGCCAGCCGCCGGATT').hamming_distance('AGG')
  end

end

File: assignments/python/point-mutations/point-mutations_test.py
try:
    from dna import DNA
except ImportError:
    raise SystemExit('Could not find dna.py. Does it exist?')

import unittest

class DNATest(unittest.TestCase):
    def test_no_difference_between_empty_strands(self):
        self.assertEqual(0, DNA('').hamming_distance(''))
  
    def test_no_difference_between_identical_strands(self):
        self.assertEqual(0, DNA('GGACTGA').hamming_distance('GGACTGA'))
  
    def test_complete_hamming_distance_in_small_strand(self):
        self.assertEqual(3, DNA('ACT').hamming_distance('GGA'))
  
    def test_hamming_distance_in_off_by_one_strand(self):
        self.assertEqual(19, DNA('GGACGGATTCTGACCTGGACTAATTTTGGGG').hamming_distance('AGGACGGATTCTGACCTGGACTAATTTTGGGG'))
  
    def test_small_hamming_distance_in_middle_somewhere(self):
        self.assertEqual(1, DNA('GGACG').hamming_distance('GGTCG'))
  
    def test_larger_distance(self):
        self.assertEqual(2, DNA('ACCAGGG').hamming_distance('ACTATGG'))
  
    def test_ignores_extra_length_on_other_strand_when_longer(self):
        self.assertEqual(3, DNA('AAACTAGGGG').hamming_distance('AGGCTAGCGGTAGGAC'))
  
    def test_ignores_extra_length_on_original_strand_when_longer(self):
        self.assertEqual(5, DNA('GACTACGGACAGGGTAGGGAAT').hamming_distance('GACATCGCACACC'))
  
    def test_does_not_actually_shorten_original_strand(self):
        self.assertEqual(1, DNA('AGACAACAGCCAGCCGCCGGATT').hamming_distance('AGGCAA'))
        self.assertEqual(4, DNA('AGACAACAGCCAGCCGCCGGATT').hamming_distance('AGACATCTTTCAGCCGCCGGATTAGGCAA'))
        self.assertEqual(1, DNA('AGACAACAGCCAGCCGCCGGATT').hamming_distance('AGG'))

if __name__ == '__main__':
    unittest.main()
